lib/IDE/Formatting.cpp

```diff
@@ -115,761 +277,2279 @@ class FormatContext {
+    auto Action = HandlePre(T, false);
+    if (Action.shouldGenerateIndentContext()) {
+      if (auto IndentCtx = getIndentContextFrom(T, Action.Trailing))
+        InnermostCtx = IndentCtx;
+    }
+    return Action.shouldVisitChildren();
+  }
+
+  bool walkToDeclPost(Decl *D) override { return HandlePost(D); }
+  bool walkToTypeReprPost(TypeRepr *T) override { return HandlePost(T); }
+
+  Stmt* walkToStmtPost(Stmt *S) override {
+    return HandlePost(S)? S : nullptr;
+  }
+
+  Expr *walkToExprPost(Expr *E) override {
+    return HandlePost(E) ? E : nullptr;
+  }
+
+  Pattern * walkToPatternPost(Pattern *P) override {
+    return HandlePost(P) ? P : nullptr;
+  }
+
+  bool shouldWalkIntoGenericParams() override { return true; }
+
+
+#pragma mark Visitation helpers
+
+  struct VisitAction {
+    enum : unsigned { Skip, VisitChildren, GetContext } action;
+    Optional<Trailing> Trailing;
+
+    bool shouldVisitChildren() const { return action >= VisitChildren; }
+    bool shouldGenerateIndentContext() const { return action >= GetContext; }
+  };
+
+  template <class T>
+  VisitAction HandlePre(T* Node, bool IsImplicit) {
+    SourceLoc Start = Node->getStartLoc(), End = Node->getEndLoc();
+
+    if (Start.isInvalid())
+      return {VisitAction::VisitChildren, None};
+
+    Optional<Trailing> Trailing = checkForTrailingTarget(End);
+    scanTokensUntil(Start);
+
+    if (!isTargetContext(Start, End) && !Trailing)
+      return {VisitAction::Skip, None};
+    if (!NodesToSkip.count(Node) && !IsImplicit)
+      return {VisitAction::GetContext, Trailing};
+    return {VisitAction::VisitChildren, Trailing};
+  }
+
+  template <typename T>
+  bool HandlePost(T* Node) {
+    return !SM.isBeforeInBuffer(TargetLocation, Node->getStartLoc());
+  }
+
+  void scanTokensUntil(SourceLoc Loc) {
+    if (InDocCommentBlock || InCommentLine)
+      return;
+    for (auto Invalid = Loc.isInvalid(); CurrentTokIt != TokenList.end() &&
+         (Invalid || SM.isBeforeInBuffer(CurrentTokIt->getLoc(), Loc));
+         CurrentTokIt++) {
+      if (CurrentTokIt->getKind() == tok::comment) {
+        CharSourceRange CommentRange = CurrentTokIt->getRange();
+        SourceLoc StartLineLoc = Lexer::getLocForStartOfLine(
+            SM, CommentRange.getStart());
+
+        // The -1 is needed in case the past-the-end position is a newline
+        // character. In that case getLocForStartOfLine returns the start of
+        // the next line.
+        SourceLoc EndLineLoc = Lexer::getLocForStartOfLine(
+            SM, CommentRange.getEnd().getAdvancedLoc(-1));
+        auto TokenStr = CurrentTokIt->getRange().str();
+        InDocCommentBlock |= SM.isBeforeInBuffer(StartLineLoc, TargetLineLoc) && !SM.isBeforeInBuffer(EndLineLoc, TargetLineLoc) &&
+            TokenStr.startswith("/*");
+        InCommentLine |= StartLineLoc == TargetLineLoc &&
+            TokenStr.startswith("//");
+      }
+    }
+  }
+
+  Optional<Trailing>
+  checkForTrailingTarget(SourceLoc End) {
```

Comment:
I found the terminology "Trailing" hard to keep straight, since we have trailing closures, trailing where clauses, trailing targets ie. "Trailing".   Would it be correct to call this something like  `PostTokenTarget`? Maybe "trailing" is the right word, but given how overloaded it is, it would be nice if we could find the most specific term here.

Reply:
The 'Token' part isn't quite right, since it can just be an empty line between two tokens. Do you mean "Trailing" specifically or "TrailingTarget" too?